### PR TITLE
[MM-16389] Enable emojis for profile icons from webhooks

### PR DESCRIPTION
**Pull request by krjn**
#### Summary
An `icon_emoji` parameter was added to incoming webhook requests, that overrides the profile icon and the `icon_url` parameter.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-16389

#### Checklist

- [x] All new/modified APIs include changes to [mattermost-redux](https://github.com/mattermost/mattermost-redux) (https://github.com/mattermost/mattermost-redux/pull/876
- [x] Has UI changes

#### Device Information
This PR was tested on:
 - Android Emulator (29)

Related PRs:
- https://github.com/mattermost/mattermost-server/pull/11586
- https://github.com/mattermost/mattermost-webapp/pull/3074
- https://github.com/mattermost/docs/pull/2832
- https://github.com/mattermost/mattermost-developer-documentation/pull/321

### CODE REVIEW

File: app/components/post_profile_picture/index.js [hunk 1]
```diff
@@ -19,12 +20,14 @@ function mapStateToProps(state, ownProps) {
     const post = ownProps.post;
     const user = getUser(state, post.user_id);
 
+    const overrideIconURL = Client4.getAbsoluteUrl(post?.props?.override_icon_url); // eslint-disable-line camelcase
```
Comment (lieut-data) [inline app/components/post_profile_picture/index.js]: Out of curiosity, is the reason this doesn't "just work" because of the execution context of RN? (I'm more familiar with React apps where a relative URL would, of course, just load normally without more configuration.)

Comment (enahum) [reply app/components/post_profile_picture/index.js]: @lieut-data relative URL's just does not work, the device needs the absolute URL as the is nothing caching the url address for you, that is why on the `Client4` object we store the server url as well as in an entity in the redux store.

File: app/components/post_profile_picture/post_profile_picture.js [hunk 2]
```diff
@@ -55,6 +55,7 @@ export default class PostProfilePicture extends PureComponent {
 
         if (fromWebHook && enablePostIconOverride) {
             const icon = overrideIconUrl ? {uri: overrideIconUrl} : webhookIcon;
+            const borderRadius = icon === webhookIcon ? ViewTypes.PROFILE_PICTURE_SIZE / 2 : 0;
```
Comment (enahum) [inline app/components/post_profile_picture/post_profile_picture.js]: why the change in the borderRadius?

Comment (krjn) [reply app/components/post_profile_picture/post_profile_picture.js]: When an emoji is shown instead of the common profile pictures, the border makes it a bit weird. After some discussion, we dropped the border for any override, emoji or not.

Comment (enahum) [reply app/components/post_profile_picture/post_profile_picture.js]: could you please share a screenshot?

Comment (krjn) [reply app/components/post_profile_picture/post_profile_picture.js]: image

(from the webapp)

Comment (hmhealey) [reply app/components/post_profile_picture/post_profile_picture.js]: Is that being done for all overridden profile pictures or just emojis though? I'd expect it to only change for emojis if that was the case

Comment (krjn) [reply app/components/post_profile_picture/post_profile_picture.js]: All overriden profile pictures.

If not done this way, we'd either have to
 - heuristically determine if it's about an emoji by checking if the URL is relative (quick, dirty way) or
 -  send the information along to the apps (it's not sent anymore - the emoji key is removed from the message props on the server before dispatching). Considering that, IIRC during some conversation we thought of changing all overriden profile pictures along the way.

If we don't do that, I'm 2/5 on changing this back and using the heuristic way. Not terribly clean, but not bad. Importantly, we could change this in the future because the emoji info is stored in the message in the database.

@aaronrothschild

Comment (enahum) [reply app/components/post_profile_picture/post_profile_picture.js]: Please run this by UX

Comment (aaronrothschild) [reply app/components/post_profile_picture/post_profile_picture.js]: @asaadmahmood I can't find Andrew Brown's username here on GH.  Adding you as a reviewer.

Quick Summary:  The purpose of this change is to provide Some better slack compatibility for developers when using our API.  This change allows a developer to override the user's profile photo and use an Emoji as the profile photo for the incoming webhook payload.  Since an emoji is often square-shaped, we made a 2/5 decision to remove the "circular mask" that frames profile photos by default (see screenshot from krjn above) when an override emoji OR photo URL is used.  

Question:   When an image is provided as the override in an incoming hook, should we "force" the 'circular mask' onto the image so it looks like a 'regular' profile picture?

IMO - I think that, Yes, we should enforce the circular mask for overridden pictures - this adds some consistency.   

@krjn I'm impartial as to *how* it is done, but obviously prefer a way that lets us change this behaviour in the future.

Comment (asaadmahmood) [reply app/components/post_profile_picture/post_profile_picture.js]: Prefer to have the circular mask, however, I'd reduce the size of the emojis so most emojis don't normally hit it. Other than that, looks good.

Comment (asaadmahmood) [reply app/components/post_profile_picture/post_profile_picture.js]: Prefer to have the circular mask, however, I'd reduce the size of the emojis so most emojis don't normally hit it. Other than that, looks good.

Comment (krjn) [reply app/components/post_profile_picture/post_profile_picture.js]: New screenshot and commit. The last message shows a case that shouldn't normally happen, but it's there to illustrate how the border is being applied for emojis. The second message is of an url-overriden profile picture.

image

Comment (andrewbrown00) [reply app/components/post_profile_picture/post_profile_picture.js]: I 0/5 that we should use circular mask. My hesitation is emojis with non-transparent backgrounds or image emojis like the last item above will have the corners clipped rather than respecting the circular mask.

To avoid the odd clipping issue, I would recommend not using the circular mask and set the dimensions (as a square) smaller (28x28) than the user avatar.

**Pro** - we avoid any odd clipping of the emojis, I'm 4/5 that I want to see the emojis as they are and not run a chance of clipping them and they loose context since we don't control what users will upload. This is less of a risk with avatars since users can crop them in the UI to their liking.
**Pro** - I'm okay with these being slightly smaller since they are non-standard avatars and have the `bot` label present.
**Con** - 0/5 if the subtle difference in avatar sizes will seem like a bug rather than a conscious decision.

**Stubbed out some options based on the feedback** - my vote is #2 (28px - no mask)
image

Comment (krjn) [reply app/components/post_profile_picture/post_profile_picture.js]: I changed it to the second option. I liked the result and after some thinking, I believe the lesser size for emojis won't look weird/buggy.

Comment (jasonblais) [reply app/components/post_profile_picture/post_profile_picture.js]: @andrewbrown00 @asaadmahmood @aaronrothschild 

1/5 I'm wondering why we need to special case it for emojis..if I have a trophy as my profile pic, it would cut off as well because of the circular border, I think? Or is the inconsistency not really a concern in this case?

Comment (aaronrothschild) [reply app/components/post_profile_picture/post_profile_picture.js]: > 1/5 I'm wondering why we need to special case it for emojis..if I have a trophy as my profile pic, it would cut off as well because of the circular border, I think? Or is the inconsistency not really a concern in this case?

@jasonblais In the case of a profile pic - you are able to choose and see the result of uploading a photo and then see how the circle gets applied, and then you could adjust the size of the trophy for example and reupload.   

In this case, we are just using an Emoji programmatically - and there is no UI to preview what it will look like.  A developer could not reasonably expect that an emoji would have a somewhat arbitrary circle applied around it IMO.

In Slack, whose functionality we are trying to replicate here on this ticket - would not apply any kind of mask to a post that uses an Emoji, so let's emulate that functionality.

Comment (krjn) [reply app/components/post_profile_picture/post_profile_picture.js]: > [...] so let's emulate that functionality.

After removing the mask, do we keep the 4px difference? (emojis being 28px instead of 32px)

Comment (jasonblais) [reply app/components/post_profile_picture/post_profile_picture.js]: @aaronrothschild Thank you! Good points raised here

Comment (aaronrothschild) [reply app/components/post_profile_picture/post_profile_picture.js]: > > [...] so let's emulate that functionality.
> 
> After removing the mask, do we keep the 4px difference? (emojis being 28px instead of 32px)

I'm going to defer to @andrewbrown00 earlier:
>To avoid the odd clipping issue, I would recommend not using the circular mask and set the dimensions (as a square) smaller (28x28) than the user avatar.  

Does that answer your question @krjn?

Comment (krjn) [reply app/components/post_profile_picture/post_profile_picture.js]: Yep. It is ready then.

File: app/components/post_profile_picture/index.js [hunk 3]
```diff
@@ -19,14 +20,20 @@ function mapStateToProps(state, ownProps) {
     const post = ownProps.post;
     const user = getUser(state, post.user_id);
 
+    // if it's relative we assume it's a link to an emoji
+    const isEmoji = post?.props?.override_icon_url?.startsWith('/') || false; // eslint-disable-line camelcase
```
Comment (hmhealey) [inline app/components/post_profile_picture/index.js]: I don't know if we can make this assumption since it's definitely possible that an integration could use a relative link for its profile image, particularly if that webhook is being used by a plugin that hosts the image itself.

Maybe when we populate the `override_icon_url`, we could just leave the `override_icon_emoji` prop on the post and use that to know for certain that this is an emoji? I think it's totally fine to have both fields since that supports both new and old apps

File: app/components/post_profile_picture/index.js [hunk 4]
```diff
@@ -19,14 +20,20 @@ function mapStateToProps(state, ownProps) {
...
+
+    const overrideIconURL = Client4.getAbsoluteUrl(post?.props?.override_icon_url); // eslint-disable-line camelcase
+
     return {
         enablePostIconOverride: config.EnablePostIconOverride === 'true' && post?.props?.use_user_icon !== 'true', // eslint-disable-line camelcase
         fromWebHook: post?.props?.from_webhook === 'true', // eslint-disable-line camelcase
         isSystemMessage: isSystemMessage(post),
         fromAutoResponder: fromAutoResponder(post),
-        overrideIconUrl: post?.props?.override_icon_url, // eslint-disable-line camelcase
+        overrideIconUrl: overrideIconURL,
```
Comment (hmhealey) [inline app/components/post_profile_picture/index.js]: Minor nit, but can you match the case for consistency?

**Comment by hmhealey** [comment]
@krjn I know my review is coming in a bit late, and since we're trying to get this in for 5.14, I'd be able to make that change to keep the override_icon_emoji prop if you think it makes sense to do it that way instead. Just let me know

File: app/components/post_profile_picture/index.js (hunk 3 above)
Comment (krjn) [reply app/components/post_profile_picture/index.js]: Agreed. Changed.

**Comment by krjn** [comment]
>Just let me know

All good :)

**Comment by hmhealey** [comment]
Thanks! Is a server PR needed to preserve that prop? 

https://github.com/mattermost/mattermost-server/blob/master/app/post_metadata.go#L66

**Comment by krjn** [comment]
Yes. Glad you asked. I pushed on https://github.com/mattermost/mattermost-webapp/pull/3074 without realizing it was merged a few hours earlier. I'll create a new one.

**Comment by hmhealey** [comment]
@jasonblais Does this look good from a PM perspective?

**Comment by jasonblais** [comment]
@hmhealey Yes, looks like Aaron and Asaad have approved

**Comment by hanzei** [comment]
@aaronrothschild @hmhealey Can this be merged?

**Comment by hanzei** [comment]
@aaronrothschild @hmhealey Can this be merged?

**Comment by aaronrothschild** [comment]
Yes, this can be merged.

File: app/components/post_profile_picture/index.js [hunk 5]
```diff
@@ -19,14 +20,17 @@ function mapStateToProps(state, ownProps) {
...
     return {
         enablePostIconOverride: config.EnablePostIconOverride === 'true' && post?.props?.use_user_icon !== 'true', // eslint-disable-line camelcase
         fromWebHook: post?.props?.from_webhook === 'true', // eslint-disable-line camelcase
         isSystemMessage: isSystemMessage(post),
         fromAutoResponder: fromAutoResponder(post),
-        overrideIconUrl: post?.props?.override_icon_url, // eslint-disable-line camelcase
+        overrideIconUrl,
         userId: post.user_id,
         isBot: (user ? user.is_bot : false),
+        isEmoji: post?.props?.override_icon_emoji?.length > 0, // eslint-disable-line camelcase
```
Comment (enahum) [inline app/components/post_profile_picture/index.js]:
```suggestion
        isEmoji: Boolean(post?.props?.override_icon_emoji), // eslint-disable-line camelcase
```

**Comment by enahum** [comment]
@aaronrothschild you always need to dev approvals before being able to merge.. I just requested a change

**Comment by hanzei** [comment]
@enahum Assad and Harrison did approve the PR. Wasn't this enough?

**Comment by enahum** [comment]
@hanzei Asaad approved on the UI/UX side ;)

**Comment by hanzei** [comment]
I see. Thought this would count as dev review.

**Comment by aaronrothschild** [comment]
@enahum Apologies - I shouldn't have said that, I had seen a previous Dev Review approval.  I should have said I approved the changes from a PM perspective already.  Thanks!

**Comment by enahum** [comment]
Just want to point out and I take some of the blame that this PR broke master as the reference for the redux changes were not included
